lib/capability/capability.go

```diff
@@ -0,0 +1,115 @@
+package capability
+
+import (
+	"fmt"
+	"strings"
+
+	configv1 "github.com/openshift/api/config/v1"
+)
+
+const (
+	CapabilityAnnotation = "capability.openshift.io/name"
+
+	DefaultCapabilitySet = configv1.ClusterVersionCapabilitySetCurrent
+)
+
+type ClusterCapabilities struct {
+	KnownCapabilities   map[configv1.ClusterVersionCapability]struct{}
+	EnabledCapabilities map[configv1.ClusterVersionCapability]struct{}
+}
+
+// SetCapabilities populates and returns cluster capabilities from ClusterVersion capabilities spec.
+func SetCapabilities(config *configv1.ClusterVersion) ClusterCapabilities {
+	var capabilities ClusterCapabilities
+	capabilities.KnownCapabilities = setKnownCapabilities(config)
+	capabilities.EnabledCapabilities = setEnabledCapabilities(config)
+	return capabilities
+}
+
+// GetCapabilitiesStatus populates and returns ClusterVersion capabilities status from given capabilities.
+func GetCapabilitiesStatus(capabilities ClusterCapabilities) configv1.ClusterVersionCapabilitiesStatus {
+	var status configv1.ClusterVersionCapabilitiesStatus
+	for k := range capabilities.EnabledCapabilities {
+		status.EnabledCapabilities = append(status.EnabledCapabilities, k)
+	}
+	for k := range capabilities.KnownCapabilities {
+		status.KnownCapabilities = append(status.KnownCapabilities, k)
+	}
+	return status
+}
+
+// CheckResourceEnablement, given resource annotations and defined cluster capabilities, checks if the capability
+// annotation exists. If so, each capability name is validated against the known set of capabilities. Each valid
+// capability is then checked if it is disabled. If any invalid capabilities are found an error is returned listing
+// all invalid capabilities. Otherwise, if any disabled capabilities are found an error is returned listing all
+// disabled capabilities.
+func CheckResourceEnablement(annotations map[string]string, capabilities ClusterCapabilities) error {
+	val, ok := annotations[CapabilityAnnotation]
+	if ok {
+		caps := strings.Split(val, "+")
+		numCaps := len(caps)
+		unknownCaps := make([]string, 0, numCaps)
+		disabledCaps := make([]string, 0, numCaps)
+
+		for _, c := range caps {
+			trimmedCap := strings.TrimSpace(c)
+			if len(trimmedCap) == 0 {
+				continue
```

Comment:
This opens up space for accepting values like `a+b++c+   +d`.  I don't think we say we'll trim spaces anywhere.  Can we drop `trimmedCap` and use the raw `c` below?

Reply:
I okay with either but figured since we didn't say they couldn't put spaces I'd check. However, I did that when we were using a coma where a space would be somewhat natural. Not the case with `+` so yeah, I'll just drop it.